File: source/ch/cyberduck/ui/cocoa/odb/WatchEditor.java
package ch.cyberduck.ui.cocoa.odb;

/*
 *  Copyright (c) 2007 David Kocher. All rights reserved.
 *  http://cyberduck.ch/
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  Bug fixes, suggestions and comments should be sent to:
 *  dkocher@cyberduck.ch
 */

import ch.cyberduck.core.Local;
import ch.cyberduck.core.Path;
import ch.cyberduck.core.TransferAction;
import ch.cyberduck.ui.cocoa.BrowserController;
import ch.cyberduck.ui.cocoa.application.NSWorkspace;
import ch.cyberduck.ui.cocoa.io.FileWatcher;
import ch.cyberduck.ui.cocoa.io.FileWatcherListener;

import org.apache.commons.lang.StringUtils;
import org.apache.log4j.Logger;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

/**
 * An editor listing for file system notifications on a particular folder
 *
 * @version $Id$
 */
public class WatchEditor extends Editor implements FileWatcherListener {
    private static Logger log = Logger.getLogger(WatchEditor.class);

    private static final Map<String, String> SUPPORTED_KQUEUE_EDITORS = new HashMap<String, String>();
    private static final Map<String, String> INSTALLED_KQUEUE_EDITORS = new HashMap<String, String>();

    static {
        SUPPORTED_KQUEUE_EDITORS.put("TextEdit", "com.apple.TextEdit");
        SUPPORTED_KQUEUE_EDITORS.put("Xcode", "com.apple.Xcode");

        Iterator<String> editorNames = SUPPORTED_KQUEUE_EDITORS.keySet().iterator();
        Iterator<String> editorIdentifiers = SUPPORTED_KQUEUE_EDITORS.values().iterator();
        while(editorNames.hasNext()) {
            String editor = editorNames.next();
            String identifier = editorIdentifiers.next();
            final String path = NSWorkspace.sharedWorkspace().absolutePathForAppBundleWithIdentifier(identifier);
            if(StringUtils.isEmpty(path)) {
                continue;
            }
            addInstalledEditor(editor, identifier);
        }
    }

    public static void addInstalledEditor(String name, String identifier) {
        SUPPORTED_KQUEUE_EDITORS.put(name, identifier);
        INSTALLED_KQUEUE_EDITORS.put(name, identifier);
    }

    public static Map<String, String> getSupportedEditors() {
        return SUPPORTED_KQUEUE_EDITORS;
    }

    public static Map<String, String> getInstalledEditors() {
        return INSTALLED_KQUEUE_EDITORS;
    }

    private FileWatcher monitor;

    /**
     * @param c
     */
    public WatchEditor(BrowserController c, Path path) {
        this(c, path.getLocal().getDefaultApplication(), path);
    }

    /**
     * @param c
     * @param bundleIdentifier
     */
    public WatchEditor(BrowserController c, String bundleIdentifier, Path path) {
        super(c, bundleIdentifier, path);
    }

    @Override
    protected TransferAction getAction() {
        return TransferAction.ACTION_RENAME;
    }

    /**
     * Edit and watch the file for changes
     */
    @Override
    public void edit() {
        NSWorkspace.sharedWorkspace().openFile(edited.getLocal().getAbsolute(),
                NSWorkspace.sharedWorkspace().absolutePathForAppBundleWithIdentifier(bundleIdentifier));
        this.watch();
    }

    /**
     * Watch the file for changes
     */
    public void watch() {
        monitor = new FileWatcher(edited.getLocal());
        monitor.register();
        monitor.addListener(this);
    }

    @Override
    protected void delete() {
        monitor.removeListener(this);
        super.delete();
    }

    @Override
    protected void setDeferredDelete(boolean deferredDelete) {
        if(!this.isOpen()) {
            this.delete();
        }
        super.setDeferredDelete(deferredDelete);
    }

    public void fileWritten(Local file) {
        log.info("fileWritten:" + file);
        this.save();
    }

    public void fileRenamed(Local file) {
        log.info("fileRenamed:" + file);
    }

    public void fileDeleted(Local file) {
        log.info("fileDeleted:" + file);
        monitor.removeListener(this);
    }
}

File: source/ch/cyberduck/ui/controller/WatchEditor.cs
// 
// Copyright (c) 2010 Yves Langisch. All rights reserved.
// http://cyberduck.ch/
// 
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// Bug fixes, suggestions and comments should be sent to:
// yves@cyberduck.ch
// 
using System;
using System.ComponentModel;
using System.Diagnostics;
using System.IO;
using Ch.Cyberduck.Core;
using org.apache.log4j;
using Path = ch.cyberduck.core.Path;

namespace Ch.Cyberduck.Ui.Controller
{
    public class WatchEditor : Editor
    {
        private static readonly Logger Log = Logger.getLogger(typeof (WatchEditor).FullName);
        private string _editor;
        private FileSystemWatcher _watcher;

        public WatchEditor(BrowserController controller, Path path, String editor)
            : base(controller, null, path)
        {
            _editor = editor;
        }

        public string Editor
        {
            get { return _editor; }
            set { _editor = value; }
        }

        protected override void edit()
        {
            Process process = new Process();
            if (Utils.IsBlank(_editor))
            {
                process.StartInfo.FileName = edited.getLocal().getAbsolute();
            }
            else
            {
                process.StartInfo.FileName = _editor;
                process.StartInfo.Arguments = "\"" + edited.getLocal().getAbsolute() + "\"";
            }
            try
            {
                process.Start();
            }
            catch (InvalidOperationException e)
            {
                Log.error(e);
                return;
            }
            catch (Win32Exception e)
            {
                Log.error(e);
                return;
            }
            _watcher = new FileSystemWatcher();
            _watcher.Path = edited.getLocal().getParent().getAbsolute();
            _watcher.Filter = edited.getLocal().getName();
            _watcher.NotifyFilter = NotifyFilters.LastAccess | NotifyFilters.LastWrite
                                    | NotifyFilters.FileName | NotifyFilters.DirectoryName;
            RegisterHandlers();
            // Begin watching.
            _watcher.EnableRaisingEvents = true;
        }

        private void FileNeedsToBeUpdated()
        {
            
        }

        private void HasRenamed(object sender, RenamedEventArgs e)
        {
            Log.debug(String.Format("HasRenamed: from {0} to {1}", e.OldFullPath, e.FullPath));
            save();
        }

        private void HasChanged(object sender, FileSystemEventArgs e)
        {
            Log.debug("HasChanged:" + e.FullPath);
            save();
        }

        protected override void delete()
        {
            _watcher.EnableRaisingEvents = false;
            RemoveHandlers();
            base.delete();
        }

        private void RegisterHandlers()
        {
            _watcher.Changed += HasChanged;
            _watcher.Renamed += HasRenamed;
        }

        private void RemoveHandlers()
        {
            _watcher.Changed -= HasChanged;
            _watcher.Renamed -= HasRenamed;
        }
    }
}